README.md
# flutter_web_scrollbar

  

A flutter web plugin for adding a customizable scrollbar to your web project.
![Scroll Bar Web](https://i.ibb.co/D7SG7KG/ezgif-com-crop.gif)
## Usage
To use this plugin add `flutter_web_scrollbar` as [a dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages).
  
Adding the scrollbar to your webpage requires a few mandatory steps: 

 1. The content of your page must be the child of a Scrollable Widget (i.e: `SingleChildScrollView` 
 2. The scrollable widget and the FlutterWebScrollbar ***must*** be children in a parent `Stack`
 3. The scrollable widget ***must*** be listed first in the `Stack`
 4. The scrollable widget must have a  `ScrollController` attached to it 
 5. The FlutterWebScrollbar must receive a callBack function that accepts a `DragUpdateDetails` as a parameter. 

**Example** 

```
import 'package:flutter/material.dart';  
import 'package:flutter_web_scrollbar/flutter_web_scrollbar.dart';  
  
class LandingPage extends StatefulWidget {  
  LandingPage({Key key}) : super(key: key);  
  @override  
  State<StatefulWidget> createState() {  
    return _LandingPageState();  
  }  
}  
  
class _LandingPageState extends State<LandingPage> {  
  ScrollController _controller;  
  
  @override  
  void initState() {  
      //Initialize the  scrollController  
      _controller = ScrollController();  
      super.initState();  
  }  
  
  void scrollCallBack(DragUpdateDetails dragUpdate) {  
    setState(() {  
      // Note: 3.5 represents the theoretical height of all my scrollable content. This number will vary for you.  
       _controller.position.moveTo(dragUpdate.globalPosition.dy * 3.5);  
    });  
  }  
  
  @override  
  Widget build(BuildContext context) {  
  
    return Scaffold(  
      backgroundColor: Colors.white,  
      body: Stack(  
        children: [  
          Container(  
            child: SingleChildScrollView(  
              //Assign the controller to my scrollable widget  
              controller: _controller,  
              child: Column(  
                children: [  
                  Container(  
                    height: MediaQuery.of(context).size.height,  
                    width: MediaQuery.of(context).size.width,  
                    color: Colors.black,  
                  ),  
                  Container(  
                    height: MediaQuery.of(context).size.height,  
                    width: MediaQuery.of(context).size.width,  
                    color: Colors.red,  
                  ),  
                  Container(  
                    height: MediaQuery.of(context).size.height,  
                    width: MediaQuery.of(context).size.width,  
                    color: Colors.green,  
                  ),  
                  Container(  
                    height: MediaQuery.of(context).size.height,  
                    width: MediaQuery.of(context).size.width,  
                    color: Colors.blue,  
                  ),  
                ],  
              ),  
            ),  
          ),  
  
  
          FlutterWebScroller(  
            //Pass a reference to the ScrollCallBack function into the scrollbar  
            scrollCallBack,  
  
            //Add optional values  
            scrollBarBackgroundColor: Colors.white,  
            scrollBarWidth: 20.0,  
            dragHandleColor: Colors.red,  
            dragHandleBorderRadius: 2.0,  
            dragHandleHeight: 40.0,  
            dragHandleWidth: 15.0,  
          ),  
        ],  
      ),  
    );  
  }  
}
````

Things to note about the above code:
- 
- Inside of our `scrollCallBack` function, we're updating the `ScrollController` that is attached to our scrollable widget. It's important to note that in the example, we are using a value of 3.5 to represent the theoretical height of our page content. The theoretical height of your page content may be different. For example, if you have to full height container as your content, you may want to use 2 rather than 3.5. Getting this value as correct as possible plays a big role in how accurate your scrolling will be. 
- Visit the repository [here](https://github.com/cairacshields/flutter_web_scrollbar_plugin) to submit issue reports and request changes
- If you'd like to help with the upkeep for the plugin, email me at `caira102797@gmail.com`

        
<a href="https://www.buymeacoffee.com/cairashields" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Pizza" style="height: 51px !important;width: 217px !important;" ></a>      


CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.

## [0.0.2] - TODO: Add release date.

* TODO: Describe initial release.

## [0.0.3] - 10/16/2020

* Fix overscrolling issue.


test/flutter_web_scrollbar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_web_scrollbar/flutter_web_scrollbar.dart';

void main() {
//  test('adds one to input values', () {
//    final calculator = Calculator();
//    expect(calculator.addOne(2), 3);
//    expect(calculator.addOne(-7), -6);
//    expect(calculator.addOne(0), 1);
//    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//  });
}


lib/flutter_web_scrollbar.dart
library flutter_web_scrollbar;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_web_plugins/flutter_web_plugins.dart';

/// Welcome to the FlutterWebScrollBar Plugin!
/// In order to assist with using the plugin directly,
/// I've set up this introduction.
///
/// Before you begin implementing this scrollbar for
/// your web project, it's important to note that the plugin is
/// still under development.
///
/// We are actively working on things like mouse scroll detection, and
/// automatically determining the height of your page content.
///
/// Please feel free to file an issue report at the repository found here:
/// https://github.com/cairacshields/flutter_web_scrollbar_plugin

class FlutterWebScrollBar {
  static void registerWith(Registrar registrar) {
    final MethodChannel channel = MethodChannel(
        'plugins.flutter.io/flutter_web_scrollbar',
        const StandardMethodCodec(),
        registrar.messenger);
    final FlutterWebScrollBar instance = FlutterWebScrollBar();
    channel.setMethodCallHandler(instance.handleMethodCall);
  }

  Future<dynamic> handleMethodCall(MethodCall call) async {
    switch (call.method) {
      case 'flutter_web_scrollbar':
        Function scrollCallBack = call.arguments['scrollCallBack'];
        Color backgroundColor = call.arguments['scrollBarBackgroundColor'];
        double barWidth = call.arguments['scrollBarWidth'];
        Color handleColor = call.arguments['dragHandleColor'];
        double handleWidth = call.arguments['dragHandleWidth'];
        double handleHeight = call.arguments['dragHandleHeight'];
        double handleBorderRadius = call.arguments['dragHandleBorderRadius'];

        return FlutterWebScroller(
          scrollCallBack,
          scrollBarBackgroundColor: backgroundColor,
          scrollBarWidth: barWidth,
          dragHandleColor: handleColor,
          dragHandleWidth: handleWidth,
          dragHandleHeight: handleHeight,
          dragHandleBorderRadius: handleBorderRadius,
        );
      default:
        throw PlatformException(
            code: 'Unimplemented',
            details: "The flutter_web_scrollbar plugin for web doesn't implement "
                "the method '${call.method}'");
    }
  }
}


class FlutterWebScroller extends StatefulWidget {

  /// Required! CallBack function used to update the scrollable content with the new drag position
  final Function scrollCallBack;

  /// Optional: Background color of the scrollbar
  final Color scrollBarBackgroundColor;

  ///Optional: Width of the scrollbar
  final double scrollBarWidth;

  ///Optional: Color of the drag handle
  final Color dragHandleColor;

  ///Optional: Width of the drag handle
  final double dragHandleWidth;

  ///Optional: Height of the drag handle
  final double dragHandleHeight;

  ///Optional: Border Radius of the drag handle
  final double dragHandleBorderRadius;

  FlutterWebScroller(this.scrollCallBack,
      {this.scrollBarBackgroundColor = Colors.black12,
      this.dragHandleColor = Colors.grey,
      this.scrollBarWidth = 20.0,
      this.dragHandleHeight = 40.0,
      this.dragHandleWidth = 15.0,
      this.dragHandleBorderRadius = 3.0});

  @override
  State<StatefulWidget> createState() {
    // TODO: implement createState
    return _FlutterWebScrollerState();
  }
}

class _FlutterWebScrollerState extends State<FlutterWebScroller> {
  static double _offset = 0;

  @override
  Widget build(BuildContext context) {
    return //Scroll bar
        Container(
            alignment: Alignment.centerRight,
            height: MediaQuery.of(context).size.height,
            width: widget.scrollBarWidth,
            margin: EdgeInsets.only(
                left:
                    MediaQuery.of(context).size.width - widget.scrollBarWidth),
            decoration: BoxDecoration(color: widget.scrollBarBackgroundColor),
            child: Container(
              alignment: Alignment.topCenter,
              child: GestureDetector(
                child: Container(
                  height: widget.dragHandleHeight,
                  width: widget.dragHandleWidth,
                  margin: EdgeInsets.only(left: 5.0, right: 5.0, top: _offset),
                  decoration: BoxDecoration(
                      color: widget.dragHandleColor,
                      borderRadius: BorderRadius.all(
                          Radius.circular(widget.dragHandleBorderRadius))),
                ),
                onVerticalDragUpdate: (dragUpdate) {
                  /// Send the new drag details to the callback in order to properly update the scrollable content position
                  widget.scrollCallBack(dragUpdate);

                  setState(() {
                    if (dragUpdate.globalPosition.dy >= 0) {
                      /// Update the offset of the drag handle to push it down or shift it up
                      _offset = dragUpdate.globalPosition.dy;


                      double maxHeight = MediaQuery.of(context).size.height - widget.dragHandleHeight;
                      _offset =(_offset>maxHeight) ? maxHeight : _offset;
                    }
                  });
                },
              ),
            ));
  }
}
